Dockerfile

```diff
@@ -3,47 +3,59 @@
 ####################################################################################################
 ################################## Stage: builder ##################################################
 
-FROM balenalib/raspberry-pi-debian-python:buster-run-20211014 as builder
-
-ENV PYTHON_DEPENDENCIES_DIR=/opt/python-dependencies
-
-RUN mkdir /tmp/build
-COPY ./ /tmp/build
-WORKDIR /tmp/build
-
+FROM balenalib/raspberry-pi-debian:buster-build-20211014 as builder
 
+# Nebra uses /opt by convention
+WORKDIR /opt/
 
+# this installs python 3.7.3 which is buster default and supported by grpcio wheel
 RUN \
     install_packages \
-            build-essential \
-            libdbus-glib-1-dev
+        python3-minimal \
+        python3-venv \
+        python3-pip \
+        build-essential \
+        libdbus-glib-1-dev
 
-RUN pip3 install --no-cache-dir --target="$PYTHON_DEPENDENCIES_DIR" .
+# This will be the path that venv uses for installation below
+ENV PATH="/opt/venv/bin:$PATH"
+
+# without pinning the pip version, we get linter warning
+COPY requirements.txt requirements.txt
+RUN python3 -m venv /opt/venv && \
+    pip install --no-cache-dir --upgrade pip==22.0.1 && \
+    pip install --no-cache-dir -r requirements.txt
 
 # firehose build, the tar is obtained from  quectel.
 # there is no install target in Makefile, doing manual copy
-RUN tar -xf quectel/qfirehose/QFirehose_Linux_Android_V1.4.9.tar
+COPY quectel quectel
+RUN tar -xf ./quectel/qfirehose/QFirehose_Linux_Android_V1.4.9.tar
 # docker linter wants WORKDIR for changing directory
-WORKDIR /tmp/build/QFirehose_Linux_Android_V1.4.9
+WORKDIR /opt/QFirehose_Linux_Android_V1.4.9
 RUN make && \
     cp QFirehose /usr/sbin/QFirehose && \
     rm -rf quectel/qfirehose
 
+WORKDIR /opt
+COPY ./ /opt
+RUN pip --no-cache-dir install .
+
 # No need to cleanup the builder
 
 ####################################################################################################
 ################################### Stage: runner ##################################################
 
-FROM balenalib/raspberry-pi-debian-python:buster-run-20211014 as runner
-
-ENV PYTHON_DEPENDENCIES_DIR=/opt/python-dependencies
+FROM balenalib/raspberry-pi-debian-python:buster-build-20211014 as runner
 
 RUN \
     install_packages \
+        python3-minimal \
```

Comment:
Surprising to see python packages installed because:
- Already using Python Buster image: raspberry-pi-debian-python.
- Using venv, so Python should be bundled with that.

Reply:
I agree with venv pulling python, I can remove that I think without any issues.

raspberrypi-debian-python image is 3.10 that creates problem for precompiled grpcio wheel. Long discussion in the abondoned [pull request](https://github.com/NebraLtd/hm-diag/pull/328)

TDLR:
* balena packages their own compiled python 3.10 linked against buster's glibc2.28
* grpcio wheels that target python 3.10, target glibc2.29. Which is generally the combo in distributions that carrry 3.10
* To use those wheels, we need python 3.7 to 3.9 which is what buster shipped. So this install
* If I don't do that, even after caching grpc compilation creates slow docker builds. 
* runner still uses the base image as buster-python because of discussion in the old PR about keeping same runner base image.

Reply:
We could always compile and cache the wheels in GitHub

Reply:
yes, that can be an option. But there will be some work there it might be easier to push nebra-raspberrypi-buster-python with upstream python.